zarr/storage.py

```diff
@@ -1907,53 +1911,65 @@ class ABSStore(MutableMapping):
     `Python Client Library <https://github.com/Azure/azure-storage-python/tree/master/azure-storage-blob>`_ version >= 1.3.0.
     """
 
-    def __init__(self, container_name, prefix, account_name, account_key):
+    def __init__(self, container, prefix, account_name=None, account_key=None, blob_service_kwargs=None):
+        self.container_name = container
+        self.prefix = normalize_storage_path(prefix)
         self.account_name = account_name
         self.account_key = account_key
-        self.container_name = container_name
-        self.prefix = normalize_storage_path(prefix)
-        self.initialize_container()
-
-    def initialize_container(self):
-        from azure.storage.blob import BlockBlobService
-        self.client = BlockBlobService(self.account_name, self.account_key)
+        if blob_service_kwargs is not None:
+             self.blob_service_kwargs = blob_service_kwargs
+        else:
+             self.blob_service_kwargs = dict()
+        self.client = BlockBlobService(self.account_name, self.account_key, **self.blob_service_kwargs)
 
     # needed for pickling
     def __getstate__(self):
         state = self.__dict__.copy()
+        del state['client']
         return state
 
     def __setstate__(self, state):
         self.__dict__.update(state)
-        self.initialize_container()
+        self.client = BlockBlobService(self.account_name, self.account_key, **self.blob_service_kwargs)
 
     def __enter__(self):
         return self
 
     def __exit__(self, *args):
         pass
 
+    @staticmethod
     def _append_path_to_prefix(path, prefix):
         return '/'.join([normalize_storage_path(prefix),
                          normalize_storage_path(path)])
 
     def full_path(self, path=None):
-        return _append_path_to_prefix(path, self.prefix)
+        return self._append_path_to_prefix(path, self.prefix)
 
     def __getitem__(self, key):
         blob_name = '/'.join([self.prefix, key])
-        blob = self.client.get_blob_to_bytes(self.container_name, blob_name)
-        if blob:
+        try:
+            blob = self.client.get_blob_to_bytes(self.container_name, blob_name)
             return blob.content
-        else:
+        except AzureMissingResourceHttpError:
             raise KeyError('Blob %s not found' % blob_name)
 
     def __setitem__(self, key, value):
+        if PY2 and isinstance(value, array.array):
+            value = value.tostring()
         blob_name = '/'.join([self.prefix, key])
-        self.client.create_blob_from_text(self.container_name, blob_name, value)
+        buffer = io.BytesIO()
+        buffer.write(value)
+        buffer.seek(0)
+        self.client.create_blob_from_bytes(self.container_name, blob_name, buffer.read())
```

Comment:
So [create_blob_from_bytes](https://azure-storage.readthedocs.io/ref/azure.storage.blob.blockblobservice.html#azure.storage.blob.blockblobservice.BlockBlobService.create_blob_from_bytes) does not accept strings, it accepts an array of bytes, unlike the [gcs equivalent](https://googlecloudplatform.github.io/google-cloud-python/latest/storage/blobs.html#google.cloud.storage.blob.Blob.upload_from_string) being used in GCSStore. Although I might be able to make this work by using different blob create functions, depending on whether input is string or bytes like.

Reply:
FWIW, looking at the source code for blockblobservice, I think this would minimise memory copies:

```python
        if PY2 and isinstance(value, array.array):
            value = value.tostring()
        blob_name = '/'.join([self.prefix, key])
        buffer = io.BytesIO(value)
        self.client.create_blob_from_stream(self.container_name, blob_name, buffer)
```

In create_blob_from_bytes() it is just creating a BytesIO and passing through to create_blob_from_stream(), so you might as well call that directly.